### PR TITLE
tests for api depth

### PR DESCRIPTION
**Pull request by scottnath**
Tests and generated content updated for the changes to api to add depth for references.

_note:_ tests are run `serial` in the api tests to avoid a knex timeout issue due to the `follow` function not using knex's `transaction` setup.

---
Relates to #408

`DCO 1.1 Signed-off-by: Scott Nath <github@scottnath.com>`

### CODE REVIEW

**Comment by scottnath** [comment]
@Snugug whoops...forgot this requirement:

mixel

File: tests/apis.js [hunk 1]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
     });
 });
 
+/**
+ * Tests attributes with references
+ * @param  {object} t - ava testing
+ * @param  {object|array} attrs object containing attributes to test
+ */
+const referencerTests = (t, attrs) => {
```
Comment (Snugug) [inline tests/apis.js]: I'd like some more comments in here explaining what each block is doing so we don't have to divine it from the code each time

File: tests/apis.js [hunk 2]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
+};
+
 //////////////////////////////
 // Utils - attributes
 //////////////////////////////
-test('Utils: attributes', t => {
-  const item = Math.round(Math.random() * (content.length - 1));
-  let expected = content[item];
+test.serial.skip('Utils: attributes - empty query', t => {
+  const random = Math.round(Math.random() * live.length - 1);
```
Comment (Snugug) [inline tests/apis.js]: This `random` was right the first time. `Math.round(Math.random() * (live.length - 1)). Need the parenthesis around the `-1` because of PEMDAS

File: tests/apis.js [hunk 3]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
+  let expected = live[random];
   if (expected === undefined) {
```
Comment (Snugug) [inline tests/apis.js]: Why does this still exist if `random` works correctly?

Comment (Snugug) [inline tests/apis.js]: Well, rather, if `random` is changed to work correctly and scope numbers from `0-(length-1)`

File: tests/apis.js [hunk 4]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
+  const attributes = apiUtils.attributes(expected.attributes, model.attributes, allTypes, query);
+
+  return attributes.then(result => {
+    t.is(typeof result, 'object', 'Should contain result, an object.');
+    t.is(Object.keys(result).length, 6, 'Should contain six main objects.');
+
+    referencerTests(t, result);
+  })
+  .catch(e => {
+    console.log(e); // eslint-disable-line no-console
```
Comment (Snugug) [inline tests/apis.js]: If you're going to catch the error, it should be `console.error` and you should `t.fail()`

File: tests/apis.js [hunk 5]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
+    expected = cloneDeep(live[(live.length - 1)]);
+  }
+
+  const model = allTypes.find(typ => {
+    return typ.id === expected['type-slug'];
+  });
+
+  // because depth is checked during attributes, depth 1 and 0 are the same for the attributes function
+  const query = {
+    depth: 0,
```
Comment (Snugug) [inline tests/apis.js]: I'm confused, this comment is not true. `depth=0` should test to make sure that references returned have their `meta` information, whereas `depth=1` should check to make sure that references returned have their full attribute information, and any references there should have their `meta` information

File: tests/apis.js (hunk 4 above)
Comment (Snugug) [inline tests/apis.js]: Applies to all caught errors

File: tests/apis.js (hunk 5 above)
Comment (Snugug) [inline tests/apis.js]: `depth=0` and not including a depth query should be equal

File: tests/apis.js [hunk 6]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
-  formatted.forEach(item => {
-    t.true(item.hasOwnProperty('id'), 'Contains ID');
-    t.true(item.hasOwnProperty('type'), 'Contains Type');
-    t.true(item.hasOwnProperty('type_slug'), 'Contains Type Slug');
-    t.true(item.hasOwnProperty('key'), 'Contains Key');
-    t.true(item.hasOwnProperty('key_slug'), 'Contains Key Slug');
-    t.true(item.hasOwnProperty('meta'), 'Contains Meta');
-    t.false(item.hasOwnProperty('attributes'), 'Does not contain attributes');
-    t.is(item.meta.url, `/api/types/${item.type_slug}/${item.id}`, 'URL points to full content item');
+test.serial.skip('Utils: Format Results - List', t => {
```
Comment (Snugug) [inline tests/apis.js]: Why is this test skipped?

File: tests/apis.js [hunk 7]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
+      t.false(item.hasOwnProperty('attributes'), 'Does not contain attributes');
+      t.is(item.meta.url, `/api/types/${item.type_slug}/${item.id}`, 'URL points to full content item');
+    });
   });
 });
 
-test('Utils: Format Results - Attributes', t => {
-  const item = Math.round(Math.random() * (content.length - 1));
-  let expected = cloneDeep(content[item]);
+test.serial.skip('Utils: Format Results - Attributes', t => {
```
Comment (Snugug) [inline tests/apis.js]: Should have a test for when there is no depth query and when there is a depth query

Comment (Snugug) [inline tests/apis.js]: And for that matter when `query` isn't passed in

File: tests/apis.js [hunk 8]
```diff
@@ -52,80 +51,300 @@ test.cb.after.always(t => {
...
-    t.true(itm.hasOwnProperty('id'), 'Contains ID');
-    t.true(itm.hasOwnProperty('type'), 'Contains Type');
-    t.true(itm.hasOwnProperty('type_slug'), 'Contains Type Slug');
-    t.true(itm.hasOwnProperty('key'), 'Contains Key');
-    t.true(itm.hasOwnProperty('key_slug'), 'Contains Key Slug');
-    t.false(itm.hasOwnProperty('meta'), 'Does not contain Meta');
-    t.true(itm.hasOwnProperty('attributes'), 'Contains attributes');
+  const formatted = apiUtils.format([expected], model.attributes, allTypes, query);
+
+  return formatted.then(result => {
```
Comment (Snugug) [inline tests/apis.js]: Going to want to check that references got populated correctly in the `format `function too

File: tests/apis.js [hunk 9]
```diff
@@ -356,17 +575,18 @@ test('APIs: Types', t => {
...
+test.serial('API: All', t => {
+  return api.all({}, allTypes).then(results => {
     t.true(results.hasOwnProperty('items'), 'Has Items');
     t.true(results.hasOwnProperty('pages'), 'Has Pagination');
     t.true(results.items.length >= generated, 'Has at least all items in it');
   });
 });
 
-test('API: All with Follow', t => {
+test.serial.skip('API: All with Follow', t => {
```
Comment (Snugug) [inline tests/apis.js]: Need a test for `follow` with no depth query (this one as it currently exists) and one where there is a `depth=1` to show they're different, and a `depth=(>1)` to show the recursion.

Comment (Snugug) [inline tests/apis.js]: Same for `API: ofType with Follow`

Comment (Snugug) [inline tests/apis.js]: And for `API: One`

Comment (Snugug) [inline tests/apis.js]: Should also include tests for when `depth` is not included in the query

File: tests/apis.js (hunk 1 above)
Comment (scottnath) [inline tests/apis.js]: moved and added comments

File: tests/apis.js [hunk 10]
```diff
@@ -36,7 +35,7 @@ test.cb.before(t => {
   });
 });
 
-test.cb.after.always(t => {
+test.cb.after(t => {
```
Comment (scottnath) [inline tests/apis.js]: this is at the bottom of the file and has `always`

File: tests/apis.js (hunk 3 above)
Comment (scottnath) [inline tests/apis.js]: changed all

File: tests/apis.js (hunk 4 above)
Comment (scottnath) [inline tests/apis.js]: changed all

File: tests/apis.js (hunk 6 above)
Comment (scottnath) [inline tests/apis.js]: this is skipped because `format` now returns a promise, which is what this test is checking for.

File: tests/apis.js (hunk 5 above)
Comment (scottnath) [inline tests/apis.js]: old comment, removed. Testing for multiple depth and no query at all

File: tests/apis.js (hunk 9 above)
Comment (scottnath) [inline tests/apis.js]: did this for `format`. I can add these, but it will add many many lines

File: tests/apis.js [hunk 11]
```diff
@@ -36,96 +35,319 @@ test.cb.before(t => {
...
+    utils.referencer(t, result);
+  })
+  .catch(e => {
+    console.error(e); // eslint-disable-line no-console
+    t.fail(e.stack);
+  });
+});
+
+test.serial.skip('Utils: attributes - depth 1', t => {
+  const random = Math.round(Math.random() * (live.length - 1));
```
Comment (Snugug) [inline tests/apis.js]: This `expected`/`model` boilerplate keeps getting reused. We should make a helper function to reduce it's overall usage

File: tests/apis.js [hunk 12]
```diff
@@ -36,96 +35,319 @@ test.cb.before(t => {
...
-    .catch(e => {
-      t.fail(e);
-    });
+  const query = {};
+
+  const attributes = apiUtils.attributes(expected.attributes, model.attributes, allTypes, query);
+
+  return attributes.then(result => {
+    t.is(typeof result, 'object', 'Should contain result, an object.');
+    t.is(Object.keys(result).length, 6, 'Should contain six main objects.');
```
Comment (Snugug) [inline tests/apis.js]: These tests will return `true` for an array of 6 items. It needs to not

Comment (Snugug) [inline tests/apis.js]: Actually, it'll return `true` for an array of 6 strings. That's not at all what we want.

File: tests/apis.js [hunk 13]
```diff
@@ -36,96 +35,319 @@ test.cb.before(t => {
...
+      utils.referencer(t, result[key]);
+    });
+  })
+  .catch(e => {
+    console.error(e); // eslint-disable-line no-console
+    t.fail(e.stack);
+  });
+});
+
+test.serial.skip('Utils: attributes - no references', t => {
```
Comment (Snugug) [inline tests/apis.js]: Why is this skipped? If there aren't any references, this should behave as before and shouldn't be skipped

File: tests/apis.js [hunk 14]
```diff
@@ -36,96 +35,316 @@ test.cb.before(t => {
...
+  });
+
+  const formatted = apiUtils.format([expected], model.attributes, allTypes, query);
+
+  return formatted.then(result => {
+    result.forEach(itm => {
+      utils.attribute(t, itm);
+      t.true(itm.hasOwnProperty('attributes'), 'Contains attributes');
+
+      Object.keys(itm.attributes).forEach(attr => {
```
Comment (Snugug) [inline tests/apis.js]: This, and the other `depth=1` version of this, feel like it can be split in to a recursive helper function that can be called as many times as we'd like instead of hard-coding this as we are

Comment (Snugug) [inline tests/apis.js]: That would greatly clean up this code and make it net easier to maintain

File: tests/apis.js (hunk 9 above)
Comment (Snugug) [inline tests/apis.js]: We need this; in the original PR this was broken, so I need tests for it to make sure it works

File: tests/apis.js [hunk 15]
```diff
@@ -442,7 +655,7 @@ test('API: ofType', t => {
   });
 });
 
-test('API: ofType with Follow', t => {
+test.serial.skip('API: ofType with Follow', t => {
```
Comment (Snugug) [inline tests/apis.js]: Same as https://github.com/punchcard-cms/punchcard/pull/519/files#r86165448

File: tests/apis.js [hunk 16]
```diff
@@ -478,28 +685,31 @@ test('API: ofType with Follow', t => {
   });
 });
 
-test('API: One', t => {
-  const item = Math.round(Math.random() * (content.length - 1));
-  let expected = cloneDeep(content[item]);
+test.serial.skip('API: One', t => {
```
Comment (Snugug) [inline tests/apis.js]: Same as https://github.com/punchcard-cms/punchcard/pull/519/files#r86165448

Comment (Snugug) [inline tests/apis.js]: In addition, this doesn't appear to actually test to make sure that we've tested for references.

File: tests/fixtures/_utils.js [hunk 17]
```diff
@@ -1,217 +1,336 @@
...
-   */
-  for (i = 0; i < t; i++) {
-    let referenced = '';
-    const name = ipsum({
-      count: 3,
-      units: 'words',
-      format: 'plain',
+  // an individual piece of content, so we need to test just the attributes
+  if (attrs.hasOwnProperty('attributes')) {
+    referencer(t, attrs.attributes);
```
Comment (Snugug) [inline tests/fixtures/_utils.js]: When using recursive functions, we want to `return` the recursive call

File: tests/fixtures/_utils.js [hunk 18]
```diff
@@ -1,217 +1,336 @@
...
+        // at this point, we've found a piece of content _inside_ a piece of content's attributes
+        if (attr[atr].hasOwnProperty('attributes')) {
+          referencer(t, attr[atr].attributes);
+
+          return;
+        }
+
+        // we've reached our depth, test `meta` exists
+        else if (attr[atr].hasOwnProperty('id')) {
+          t.true(attr[atr].hasOwnProperty('meta'), 'attribute in array contains meta');
```
Comment (Snugug) [inline tests/fixtures/_utils.js]: Test that `meta` also has `url` inside it, and all of the goodies that go in to that

File: tests/fixtures/_utils.js [hunk 19]
```diff
@@ -1,217 +1,336 @@
...
+      const item = attrs[attr];
+
+      // if attributes exists, we're still digging down in our depth, recurse!
+      if (item.hasOwnProperty('attributes')) {
+        referencer(t, item.attributes);
+
+        return;
+      }
+
+      // if it's an array, we're in a repeatable - recurseit!
```
Comment (Snugug) [inline tests/fixtures/_utils.js]: Be consistent with how you joke in your comments. Need a consistent spelling of `recurseit`. Either `recurse-it`, `recurseit`, or `recursit`. Choose one and stick with it.

File: tests/fixtures/_utils.js [hunk 20]
```diff
@@ -1,217 +1,336 @@
...
 
-    const sunrise = moment().format('hh:mm');
-    const sunset = moment().format('hh:mm');
+          return;
+        }
 
-    const type = types[Math.round(Math.random() * 4)];
+        // no attributes - it should have a meta then
+        Object.keys(item).forEach(itm => {
+          t.true(item[itm].hasOwnProperty('meta'), 'attribute in array contains meta');
```
Comment (Snugug) [inline tests/fixtures/_utils.js]: All of these `meta` checks need `url` checks to go with

File: tests/fixtures/_utils.js [hunk 21]
```diff
@@ -1,217 +1,336 @@
...
-      'attributes': values,
-      'key': name,
-      'key-slug': slugify(name),
-      type,
-      'type-slug': slugify(type),
-      'revision': Math.round(Math.random() * 40),
-    };
-
-    lives[i] = live;
+const attribute = (t, attr) => {
```
Comment (Snugug) [inline tests/fixtures/_utils.js]: This needs a better name. Its name suggests checking attributes, where its checking formatted.

File: tests/apis.js [hunk 22]
```diff
@@ -36,96 +35,319 @@ test.cb.before(t => {
...
 
-test.cb.after.always(t => {
-  const items = types.map(type => {
-    return database('live').where('type', type).del().then(() => {
-      return database('schedule').where('type', type).del();
-    });
+//////////////////////////////
+// Utils - attributes
+//////////////////////////////
+test.serial.skip('Utils: attributes - empty query', t => {
```
Comment (scottnath) [inline tests/apis.js]: nope. new attributes returns a promise

File: tests/apis.js (hunk 12 above)
Comment (scottnath) [inline tests/apis.js]: That is not the _only_ test going on here. That's just there to make sure the tests fail if any number other than six elements are in that array. The tests continue in `utils.referencer` which is a large, recursing function which tests at great depth.

File: tests/apis.js [hunk 23]
```diff
@@ -36,96 +35,319 @@ test.cb.before(t => {
...
+  });
 });
 
-//////////////////////////////
-// Utils - attributes
-//////////////////////////////
-test('Utils: attributes', t => {
-  const item = Math.round(Math.random() * (content.length - 1));
-  let expected = content[item];
+test.serial.skip('Utils: attributes - depth 0', t => {
```
Comment (scottnath) [inline tests/apis.js]: nope, returns a promise now - thus the `return attributes.then(result => {` below

File: tests/apis.js (hunk 13 above)
Comment (scottnath) [inline tests/apis.js]: we worked last week together to add promises to most of these functions. These tests reflect the new promise interface.

File: tests/apis.js (hunk 9 above)
Comment (scottnath) [inline tests/apis.js]: how about I add them in the actual PR for the code?

File: tests/apis.js (hunk 15 above)
Comment (scottnath) [inline tests/apis.js]: yup...let me add in actual PR please

File: tests/apis.js (hunk 11 above)
Comment (scottnath) [inline tests/apis.js]: excellent! created a `testables` function for this. reduced the number of lines!

File: tests/fixtures/_utils.js (hunk 17 above)
Comment (scottnath) [inline tests/fixtures/_utils.js]: fixed in whole function

File: tests/fixtures/_utils.js (hunk 18 above)
Comment (scottnath) [inline tests/fixtures/_utils.js]: now checks that url exists

File: tests/apis.js [hunk 24]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
-    return database('live').where('type', type).del().then(() => {
-      return database('schedule').where('type', type).del();
-    });
+/**
+ * Randomly selects a piece of content from source, and it's content-type model
+ *
+ * @param  {object} source object of content
+ * @returns {object}  - selected content and its model
+ */
+const testables = source => {
```
Comment (Snugug) [inline tests/apis.js]: I'd like to see this in the `fixtures` file instead of hanging out at the top of API tests. This way, we can reuse this all over the place (like in schedule)

File: tests/apis.js [hunk 25]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
+  const attributes = apiUtils.attributes(testable.expected.attributes, testable.model.attributes, allTypes, query);
+
+  return attributes.then(result => {
+    t.is(typeof result, 'object', 'Should contain result, an object.');
+    t.is(Object.keys(result).length, 6, 'Should contain six main objects.');
+
+    utils.referencer(t, result);
+  })
+  .catch(e => {
+    console.log(e); // eslint-disable-line no-console
```
Comment (Snugug) [inline tests/apis.js]: Why did you remove the `t.fail(e.stack)`?

File: tests/apis.js [hunk 26]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
+
+  const attributes = apiUtils.attributes(testable.expected.attributes, testable.model.attributes, allTypes, query);
+
+  return attributes.then(result => {
+    t.is(typeof result, 'object', 'Should contain result, an object.');
+
+    utils.referencer(t, result);
+  })
+  .catch(e => {
+    console.log(e); // eslint-disable-line no-console
```
Comment (Snugug) [inline tests/apis.js]: Why did you remove the `t.fail(e.stack)`?

File: tests/apis.js [hunk 27]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
```
Comment (Snugug) [inline tests/apis.js]: Why did you remove the `t.fail(e.stack)`?

File: tests/apis.js [hunk 28]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
+
+  return attributes.then(result => {
+    t.is(typeof result, 'object', 'Should contain result, an object.');
+
+    Object.keys(result).forEach(key => {
+      utils.referencer(t, result[key]);
+    });
+  })
+  .catch(e => {
+    console.log(e); // eslint-disable-line no-console
```
Comment (Snugug) [inline tests/apis.js]: Why did you remove the `t.fail(e.stack)`?

File: tests/apis.js (hunk 9 above)
Comment (Snugug) [inline tests/apis.js]: Why not add them here? They're needed, and this is the PR for adding in the tests. As it currently stands, I can't guarantee that the tests that worked previously (and aren't touched by the upcoming changes) aren't broken.

File: tests/apis.js [hunk 29]
```diff
@@ -478,28 +626,23 @@ test('API: ofType with Follow', t => {
...
-  let expected = cloneDeep(content[item]);
-
-  if (expected === undefined) {
-    expected = cloneDeep(content[(content.length - 1)]);
-  }
 
-  const model = allTypes.find(typ => {
-    return typ.id === expected['type-slug'];
-  });
+test.serial.skip('API: One', t => {
```
Comment (Snugug) [inline tests/apis.js]: https://github.com/punchcard-cms/punchcard/pull/519#discussion_r86224422

File: tests/apis.js [hunk 30]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
+  };
+
+  const formatted = apiUtils.format([testable.expected], testable.model.attributes, allTypes, query);
+
+  return formatted.then(result => {
+    result.forEach(itm => {
+      utils.formatted(t, itm);
+      t.true(itm.hasOwnProperty('attributes'), 'Contains attributes');
+
+      Object.keys(itm.attributes).forEach(attr => {
```
Comment (Snugug) [inline tests/apis.js]: https://github.com/punchcard-cms/punchcard/pull/519#discussion_r86223908

File: tests/apis.js [hunk 31]
```diff
@@ -36,96 +34,258 @@ test.cb.before(t => {
...
```
Comment (Snugug) [inline tests/apis.js]: https://github.com/punchcard-cms/punchcard/pull/519#discussion_r86223908

File: tests/apis.js (hunk 12 above)
Comment (Snugug) [inline tests/apis.js]: The fact that it goes on to test with `utils.referencer` doesn't negate the fact that these tests can easily result in false positives

**Comment by scottnath** [comment]
@Snugug updated and ready for review

File: tests/apis.js [hunk 32]
```diff
@@ -526,32 +470,201 @@ test.serial('APIs: Types', t => {
...
+
+    t.true(results.hasOwnProperty('items'), 'Has Items');
+    t.true(results.hasOwnProperty('pages'), 'Has Pagination');
+    t.true(results.items.length >= generated, 'Has at least all items in it');
+
+    results.items.forEach(item => {
+      // Ignore empty object
+      if (Object.keys(item).length !== 0) {
+        one = true;
+        utils.formatted(t, item);
```
Comment (Snugug) [inline tests/apis.js]: For instance, for all of these `APIs - all` tests _without_ `follow`, they all should return _identical_ results, because `depth` should only affect this if `follow` is called and the actual content gets generated. It's impossible, without me introspecting each test at run-time, to know if these tests are passing because they are producing the correct results, or passing because of a false-positive from the shared `utils.formatted`

File: tests/apis.js [hunk 33]
```diff
@@ -526,32 +470,201 @@ test.serial('APIs: Types', t => {
...
+        t.true(item.hasOwnProperty('attributes'), 'Contains attributes');
```
Comment (Snugug) [inline tests/apis.js]: Without `follow`, these should not have attributes. It should be the listing of items, with `meta`. `depth` does not mean anything _unless_ the actual content is being pulled from the DB. Listing endpoints _should not do that without `follow`_

Comment (Snugug) [inline tests/apis.js]: The _only_ endpoint where `depth` should have an affect _without_ `follow` is `apis.one`

File: tests/apis.js [hunk 34]
```diff
@@ -526,32 +470,201 @@ test.serial('APIs: Types', t => {
...
```
Comment (Snugug) [inline tests/apis.js]: Same as `depth=1` test above

File: tests/apis.js [hunk 35]
```diff
@@ -526,32 +470,201 @@ test.serial('APIs: Types', t => {
...
+
+    t.true(results.hasOwnProperty('items'), 'Has Items');
+    t.true(results.hasOwnProperty('pages'), 'Has Pagination');
+    t.true(results.items.length >= generated, 'Has at least all items in it');
+
+    results.items.forEach(item => {
+      // Ignore empty object
+      if (Object.keys(item).length !== 0) {
+        one = true;
+        utils.formatted(t, item);
```
Comment (Snugug) [inline tests/apis.js]: Same ambiguity

File: tests/apis.js [hunk 36]
```diff
@@ -526,32 +470,201 @@ test.serial('APIs: Types', t => {
...
```
Comment (Snugug) [inline tests/apis.js]: Same ambiguity

File: tests/apis.js [hunk 37]
```diff
@@ -593,31 +706,123 @@ test.serial('API: Content - Descending', t => {
...
+  return api.ofType({
+    depth: 0,
+  }, testable.model, allTypes).then(results => {
+    let one = false;
+
+    results.items.forEach(itm => {
+      // Ignore empty object
+      if (Object.keys(itm).length !== 0) {
+        one = true;
+        utils.formatted(t, itm);
```
Comment (Snugug) [inline tests/apis.js]: Same comments as `API: all` tests

**Comment by scottnath** [comment]
@Snugug changes added. Depth and follow _mostly_ recurse in [this section](https://github.com/punchcard-cms/punchcard/pull/519/commits/e9f4a952a437a0021e31907686d964893aa611b7#diff-081284283518838085862d8ff31d2addR241)

This should cover `follow` being the only way to not have `meta` and instead have `attributes`.

good luck!

**Comment by scottnath** [comment]
# API: All

## no follow, no depth (current)

[example](https://moriarty-blue.herokuapp.com/api)

`api.all{}`

returns:

```
{
  "items": [
    ...
    {
      "id": "3b201321-0a33-42d0-991a-efb1ba4a2599",
      "type": "Demo",
      "type_slug": "demo",
      "key": "Car Dashboard Demo",
      "key_slug": "car-dashboard-demo",
      "meta": {
        "url": "/api/types/demo/3b201321-0a33-42d0-991a-efb1ba4a2599"
      }
    },
    ...
  ],
}
```

## follow, no depth (current)

[example](https://moriarty-blue.herokuapp.com/api?follow=1)

`api.all{{follow: true}}`

returns:

```
{
  "items": [
    ...
    {
      "id": "3b201321-0a33-42d0-991a-efb1ba4a2599",
      "type": {
        "name": "Demo",
        "slug": "demo",
        "url": "/api/types/demo"
      },
      "key": "Car Dashboard Demo",
      "key_slug": "car-dashboard-demo",
      "attributes": {
        "title": "Car Dashboard Demo",
        "short-description": "Imagine you're in the driver's seat and Watson is your co-pilot. Watson can understand your entries and respond accordingly.",
        "video-url": "https://github.com/watson-developer-cloud/car-dashboard",
        "taxonomy-use-case": null,
        "demo-ui-link": "https://conversation-demo.mybluemix.net/",
        "source-link": "https://github.com/watson-developer-cloud/car-dashboard",
        "api": [
          "26fef183-1b8c-4257-9d0e-37009ffea79b"
        ],
      }
    },
    ...
  ],
}
```

## follow, no depth (proposed in #408)

[example](https://moriarty-blue.herokuapp.com/api?follow=1)

`api.all{{follow: true}}`

returns:

```
{
  "items": [
    ...
    {
      "id": "3b201321-0a33-42d0-991a-efb1ba4a2599",
      "type": {
        "name": "Demo",
        "slug": "demo",
        "url": "/api/types/demo"
      },
      "key": "Car Dashboard Demo",
      "key_slug": "car-dashboard-demo",
      "attributes": {
        "title": "Car Dashboard Demo",
        "short-description": "Imagine you're in the driver's seat and Watson is your co-pilot. Watson can understand your entries and respond accordingly.",
        "video-url": "https://github.com/watson-developer-cloud/car-dashboard",
        "taxonomy-use-case": null,
        "demo-ui-link": "https://conversation-demo.mybluemix.net/",
        "source-link": "https://github.com/watson-developer-cloud/car-dashboard",
        "api": [
          {
            "id": "26fef183-1b8c-4257-9d0e-37009ffea79b",
            "type": "Demo",
            "type_slug": "demo",
            "key": "Conversation",
            "key_slug": "conversation",
            "meta": { // no depth = no attributes in references
              "url": "/api/types/api/26fef183-1b8c-4257-9d0e-37009ffea79b"
            }
          },
        ],
      }
    },
    ...
  ],
}
```

## no follow, with depth (proposed in #408)

[example](https://moriarty-blue.herokuapp.com/api?depth=1)

`api.all{{depth: 1}}`

returns:

```
{
  "items": [
    ...
    {
      "id": "3b201321-0a33-42d0-991a-efb1ba4a2599",
      "type": "Demo",
      "type_slug": "demo",
      "key": "Car Dashboard Demo",
      "key_slug": "car-dashboard-demo",
      "meta": { // if no follow === no attributes
        "url": "/api/types/demo/3b201321-0a33-42d0-991a-efb1ba4a2599"
      }
      
    },
    ...
  ],
}
```
